server/http/restapi/kv.go

```diff
@@ -0,0 +1,87 @@
+package restapi
+
+import (
+	// nolint:gosec
+
+	"encoding/json"
+	"net/http"
+
+	"github.com/gorilla/mux"
+	"github.com/pkg/errors"
+
+	"github.com/mattermost/mattermost-plugin-apps/server/utils/httputils"
+)
+
+// TODO use raw byte API: for now all JSON is re-encoded to use api.Mattermost API
+
+func (a *restapi) kvList(w http.ResponseWriter, req *http.Request, botUserID, prefix string) {
+	// <><>TODO kvList
+}
+
+func (a *restapi) kvGet(w http.ResponseWriter, req *http.Request, botUserID, prefix string) {
+	id := mux.Vars(req)["key"]
+	out := map[string]interface{}{}
+	err := a.api.AppServices.KVGet(botUserID, prefix, id, out)
+	if err != nil {
+		httputils.WriteInternalServerError(w, err)
+		return
+	}
+	httputils.WriteJSON(w, out)
+}
+
+func (a *restapi) kvHead(w http.ResponseWriter, req *http.Request, botUserID, prefix string) {
+	// TODO "HEAD"
+}
+
+func (a *restapi) kvPut(w http.ResponseWriter, req *http.Request, botUserID, prefix string) {
+	id := mux.Vars(req)["key"]
+
+	in := map[string]interface{}{}
+	// TODO size limit
+	err := json.NewDecoder(req.Body).Decode(&in)
+	if err != nil {
+		httputils.WriteBadRequestError(w, err)
+		return
+	}
+
+	// <><>TODO atomic support
+	// <><>TODO TTL support
+	changed, err := a.api.AppServices.KVSet(botUserID, prefix, id, in)
+	if err != nil {
+		httputils.WriteInternalServerError(w, err)
+		return
+	}
+	httputils.WriteJSON(w, map[string]interface{}{
+		"changed": changed,
+	})
+}
+
+func (a *restapi) kvDelete(w http.ResponseWriter, req *http.Request, botUserID, prefix string) {
+	id := mux.Vars(req)["key"]
+	err := a.api.AppServices.KVDelete(botUserID, prefix, id)
+	if err != nil {
+		httputils.WriteInternalServerError(w, err)
+		return
+	}
+}
+
+func (a *restapi) handleKV(
```

Comment:
Is this called by the app? How can we sandbox one app KV store from another? One user's data from another? probably JWTs should be checked here? Or can we incorporate OAuth2 scopes here?

Reply:
> How can we sandbox one app KV store from another? 

By checking the userID to be that of the Bot, and namespacing accordingly. No user-level isolation inside the app (expect the OAuth2 tokens, coming soon). Similar to current plugins.

Reply:
All KV operations must be performed (for now) with the Bot token. We _could_ partition by the actual userID, but that'd be a bit more complexity for the App to keep track of. I really hope that we figure a better state storage solution soon, but for 1.0 this should be suffiicient?

Reply:
No user partition is fine at this point. But we should maintain data security for customers. What prevents Malicious App to send first customer's userID, get their data and transfer it to the second customer?